ota_proxy/db.py

```diff
@@ -105,84 +76,180 @@ def _connect_db(self, init: bool):
+                    (datetime.now().timestamp(), res.url),
+                )
+
+                return res
+            else:
+                return
+
+    def lookup_all(self) -> List[CacheMeta]:
+        with self._con as con:
+            cur = con.execute(f"SELECT * FROM {self.TABLE_NAME}", ())
+            return [CacheMeta.row_to_meta(row) for row in cur.fetchall()]
+
+    def rotate_cache(self, bucket: int, num: int) -> Union[str, None]:
+        """Rotate cache entries in LRU flavour.
+
+        Args:
+            bucket: which bucket for space reserving
+            num: num of entries needed to be deleted in this bucket
+
+        Return:
+            A list of hashes that needed to be deleted for space reserving,
+                or None if no enough entries for space reserving.
+        """
+        # first, check whether we have required number of entries in the bucket
+        with self._con as con:
+            cur = con.execute(
+                f"SELECT COUNT(*) FROM {self.TABLE_NAME} WHERE bucket=? ORDER BY last_access LIMIT ?",
+                (bucket, num),
+            )
+            _count = cur.fetchone()[0]
+
+            # NOTE: if we can upgrade to sqlite3 >= 3.35,
+            # use RETURNING clause instead of using 2 queries as below
+
+            # if we have enough entries for space reserving
+            if _count >= num:
+                # first select those entries
+                cur = con.execute(
+                    (
+                        f"SELECT * FROM {self.TABLE_NAME} "
+                        "WHERE bucket=? "
+                        "ORDER BY last_access "
+                        "LIMIT ?"
+                    ),
+                    (bucket, num),
+                )
+                _rows = cur.fetchall()
+
+                # and then delete those entries with same conditions
+                con.execute(
+                    (
+                        f"DELETE FROM {self.TABLE_NAME} "
+                        "WHERE bucket=? "
+                        "ORDER BY last_access "
+                        "LIMIT ?"
+                    ),
+                    (bucket, num),
+                )
+
+                return [row["hash"] for row in _rows]
+
+
+def _proxy_wrapper(attr_n):
+    def _wrapped(self, *args, **kwargs):
+        # get the handler from underlaying db connector
+        def _inner():
+            _db = self._thread_local.db
+            f = partial(getattr(_db, attr_n), *args, **kwargs)
+            return f()
+
+        # inner is dispatched to the db connection threadpool
+        fut = self._executor.submit(_inner)
+        return fut.result()
+
+    return _wrapped
+
+
+def _proxy_cls_factory(cls, *, target):
+    for attr_n, attr in target.__dict__.items():
+        if not attr_n.startswith("_") and callable(attr):
+            # override method
+            setattr(cls, attr_n, _proxy_wrapper(attr_n))
+
+    return cls
+
+
+@partial(_proxy_cls_factory, target=OTACacheDB)
```

Comment:
@Bodong-Yang Don't you think this code is difficult? If there is no other implementation it is OK.

Reply:
Emmm there are straight forward way to implement proxy class by coping-pasting every methods from the proxy target class, but that will make the code hard to be maintained(when original class is updated, the proxy class also needs to be updated).

So I use this generic way, which doesn't care about the details of the proxy target.